### ISSUE TITLE
Status of Implementation of Sub-buffers?

### ISSUE THREAD

**Issue by jon-chuang** [illuhad/hipSYCL#193]
What is the status of the implementation of sub buffers?
As seen pg. 94 SYCL spec (https://www.khronos.org/registry/SYCL/specs/sycl-1.2.1.pdf)
```
buffer<int,2> parent_buffer { range<2>{ 8,8 } }; // Create 2-d buffer with 8x8 ints
buffer<int,2> sub_buf1 { parent_buffer, /*offset*/ range<2>{ 2,0 }, /*size*/ range<2>{ 2,8 } };
```

My usage:
```
sycl::buffer<uint64_t> sub_buf_0{buf_operand, sycl::range<1>(k), sycl::range<1>(num_threads)};
```
throws error
`cl::sycl::buffer<unsigned long, 1, std::allocator<unsigned long> >::buffer(cl::sycl::buffer<unsigned long, 1, std::allocator<unsigned long> >, cl::sycl::id<1> const&, cl::sycl::range<1> const&)`

**Comment by jon-chuang** [comment]
It seems it is unimplemented. Could someone guide me through the quick and dirty way to implement it? Maybe I will end up contributing something.

**Comment by jon-chuang** [comment]
From what I understand, all that needs to be implemented are some checks that there are no overlaps with other subbuffers, check that a subbuffer will not be a subbuffer of another subbuffer, is within bounds, is contiguous in the context of dim > 1. Actually not even sure the first check is required, but seems pretty necessary.

**Comment by illuhad** [comment]
Yes it is unimplemented (I will address your PR separately, don't worry :) ).

The main issue is that currently it cannot be efficiently implemented, so any attempt to implement can only be a dummy implementation.
The reason for this is that both the algorithms and data structures are lacking mechanisms to handle data transfers and data dependencies at a subbuffer granularity, so the subbuffer will have to transfer the data of the entire buffer nonetheless.
There's currently a major development effort to replace pretty much the entire hipSYCL runtime with a new one (the `scheduling-rework` branch). This effort will bring many new features, one of which will also be data management at subbuffer granularity.

**Comment by jon-chuang** [comment]
Hi @illuhad , does this mean that there's currently no way of reading and writing to global memory from a shared memory at a finer granularity than buffers? How about atomics? That sounds interesting, how can I find out more about this work? I know very little about the runtime. How can I see some documentation of proposed changes and current features? Just for my curiosity.

I was just hacking around with the pull request, I guess I will just comment out my current code utilising subbuffers and go for a crappier implementation in the meantime.

**Comment by illuhad** [comment]
You can read and write at smaller granularity in your kernels or host accessors, but data transfers for accessors and kernel dependency calculation are always done for the entire buffer.
Same for atomics; inside your kernel you can perform atomic operations on whatever you want, but if the atomic memory location comes from a buffer accessor, the entire buffer has been transferred to your device.

The explicit memory copy functions are implemented such that they can copy at smaller granularity, if I remember correctly.

The new runtime is barely documented at this point because it's a massive redesign (for example, it also generalizes the entire backend interface, allowing us to hook into more backend APIs in the future, such as e.g. targeted OpenMP) and the design was still in flux for a long time. Now that many aspects of the overall design have more or less stabilized, I had planned to document it and describe how it works in a bit more detail because it will be much more complex and sophisticated than we currently have. It will still take some time until it is ready, it's not even testable at the moment.

If you want to already know more about some specific aspects of it, just open a new issue :)

**Comment by jon-chuang** [comment]
I am only interested in accessing smaller levels of granularity on the device for now. That's because I have a complex data operation that involves many data interdependencies among different kernels. Is this still achievable with what we have? If I can get this done within the afternoon, it might still be a good learning experience for me, also to understand the library.

But actually, what you seem to be saying is that I do not have to be worried about on-device granularity, as this is handed by the devices cache management system, is this correct?

**Comment by jon-chuang** [comment]
Actually, the main thing is that I know how the data accesses are mutually exclusive, but if I pass a buffer with `read_write` permissions, that becomes blocking for subsequent kernels put into a queue. So I would have to either rely on subbuffers or atomics. Or, I could copy into new buffers and copy back at the end.

**Comment by illuhad** [comment]
Not sure I understand what you mean. Of course, if you have two kernels requiring the same buffer with any sort of write access mode (including `read_write`), that constitutes a dependency between the kernels according to the SYCL spec, which makes sense. I don't see how atomics or subbuffers would help?

**Comment by jon-chuang** [comment]
[C] = {[A][B]}
C is split into subbuffers A, B
q(ker_0(A), ker_1(B))
Now I can operated on them independently. etc.

**Comment by illuhad** [comment]
Okay, I see. So, the situation is that `ker_0` and `ker_1` are too small to utilize the GPU well, so you want to run multiple kernels simultaneously to improve device utilization.
In the current implementation you will also have to create two queues to run kernels concurrently, as each queue wraps one stream, and CUDA will require multiple streams for concurrent kernels.
The new runtime will automatically create as many streams as there are DMA engines and kernel concurrency capability on the device, and will automatically distribute operations across the streams. So, in the future, a single queue is sufficient.

Note that in SYCL 1.2.1, for this you do not even need subbuffers, you could simply create a ranged accessor.

To remove the dependency, you would need to modify (I think in `buffer.cpp`) the dependency calculation and also store access ranges in the access tracker.

If it's just a single work group, it may be easiest to just merge those two kernels into one with some branching depending on work group?

**Comment by jon-chuang** [comment]
Branching is not an option, I may have 16 or 64 workgroups required.

What is the status of the implementation of ranged access? Is it going to be part of the new runtime?

**Comment by jon-chuang** [comment]
I am quite curious and interested to contribute to the implementation of the said asynchronous graph.

**Comment by illuhad** [comment]
Yes (it boils down to the same mechanisms underneath as subbuffers, but is usually much easier to work with for the user).

Another work around for you might be cuda unified memory. It's not particularly well tested, so might need some improvements, but there's some unified memory support. Passing the buffer property `sycl::hipsycl::property::bufer::use_svm` at construction should instruct the buffer to allocate unified memory.
You could then obtain the unified memory pointer from the host accessor. Then submit kernel without any accessors and do manual waiting by calling `wait()` on the queues or the events returned from `submit()`.
If you want compatibility with Intel USM, you could just create a simple `sycl::malloc_shared()` function that calls `cudaMallocManaged()` underneath

**Comment by jon-chuang** [comment]
I may fiddle with UM later. For now, I will focus on getting my implementation working. Could you give a a reply to the rest of my comments?

**Comment by illuhad** [comment]
>  The worst thing is that one cannot synchronise across the entire job, only a workgroup. 

SYCL follows the same synchronization model as OpenCL.

> (say if I had more tasks, then some tasks with no dependencies could not jump ahead of the queue, if this is how I understand CUDA streams?). How do I get out-of-order scheduling?

CUDA streams are in-order, but SYCL requires out-of-order behavior from queues. So, at some point we need to serialize operations to the stream. But, until that moment the SYCL runtime can reorder operations. The current runtime works with cuda stream callbacks that, when triggered cause the SYCL runtime to decide on demand which tasks to submit next. So, you will get out-of-order behavior.
The main issue is that for short kernels there's some overhead which is why the new runtime will switched to a batched submission model with ahead-of-time scheduling

> I didn't realise we were using streams, as I thought one of the promises of SYCL was the use of an asynchronous task graph that calculates data dependencies and schedules kernels automatically. However, this still seems to be in the pipelines.

Well CUDA is a low level API, and since we build on CUDA directly we have to serialize operations to streams at some point. An interesting alternative that I want to play around with in the future is CUDA graphs, but AMD does not yet support it in HIP, so we still need streams at least for the AMD side.

>  In particular, I need to know if get_access is blocking across two separate queues.

`get_access` itself never blocks, but SYCL has to prevent race conditions, so if two kernels have conflicting accesses, SYCL needs to order them, regardless of the queue or context from which they come. SYCL relies on a global task graph.

> the term queue is not the most accurate, as it is intended to be an out-of-order queue? It is more like a pool.

yes, I am very unhappy with this term myself. It's really just a mechanism to submit tasks to a global task graph and to automatically attach some information to those tasks (i.e., on which device to execute).

**Comment by jon-chuang** [comment]
I think it should be called graph. So graph.submit(..) etc.

**Comment by illuhad** [comment]
Yes, I agree... I think it comes from the time when SYCL was very closely tied to OpenCL and you could just map this to OpenCL out of order queues..

**Comment by illuhad** [comment]
> I am quite curious and interested to contribute to the implementation of the said asynchronous graph.

That's awesome, I would love some help :) Let me think about how to best get started and then get back to you, okay?

**Comment by jon-chuang** [comment]
Hi okay, sure. If I ever feel like toying with UI apps, I may also look into the task graph visualiser, sort of like what tensorboard has. Perhaps it could like to NVIDIA like the nsight systems or compute, and show a playback of kernel execution. Perhaps, meta data information like which SM it was executing on can also be fed in and it will colour the kernel node differently. It's a bit of a dream, but good to dream.

**Comment by jon-chuang** [comment]
Hi, how do I tell if multiple kernels are being run? Even though I place my operations in separate queues, it seems they are running in the same stream. Will utilising 
`vector<sycl::buffer<T>>` 
allow me to delegate different buffers in a single vector to different streams? Furthermore, I am experiencing some memory bugs. 

Here is my nsight systems timeline showing that only a single stream is running :(

![Screenshot from 2020-01-17 06-06-25](https://user-images.githubusercontent.com/9093549/72588340-df205a00-38ef-11ea-81e1-1ff7bb5fe060.png)

**Comment by DieGoldeneEnte** [comment]
Have you checked, if your kernels are small enough (in terms of resource usage like local memory) to run in parallel?

**Comment by illuhad** [comment]
There's no way to tell if kernels are run concurrently from the point of view of SYCL, since this is highly backend (in this case, CUDA) and hardware specific.
As @DieGoldeneEnte points out, resource usage can prevent kernels running concurrently. Local memory, registers, number of warps etc. Additionally, there are GPUs that may not support concurrent kernel execution for market segmentation reasons. I'm not sure about the current situation, but if I remember correctly support for that used to be severely restricted in consumer GeForce GPUs.

**Comment by illuhad** [comment]
> Will utilising `vector<sycl::buffer<T>>` allow me to delegate different buffers in a single vector to different streams? 

Not sure what you mean. A buffer is not bound to a stream. A kernel is.

> Furthermore, I am experiencing some memory bugs.

Please provide a minimal program that reproduces the issue...
